File: string/longest-valid-parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        last = 0
        n = 0
        res = 0
        for i, c in enumerate(s):
            if c == "(": 
                n += 1
            else:
                n -= 1
                if n < 0:
                    last = i + 1
                    n = 0
            if n == 0:
                res = max(res, i - last + 1)
        res = max(res, len(s) - last - n)
        return res

File: string/longest-valid-parentheses.rs
use std::cmp::max;

impl Solution {
    pub fn longest_valid_parentheses(s: String) -> i32 {
        let mut stack = Vec::new();
        stack.push(-1);
        let mut n = 0;
        for (i, c) in s.chars().enumerate() {
            match c {
                '(' => stack.push(i as i32),
                _ => {
                    stack.pop();
                    match stack.last() {
                        None => stack.push(i as i32),
                        Some(j) => n = max(n, i as i32 - j),
                    }
                }
            }
        }
        n
    }
}
